In [1]:
import numpy as np 
import pandas as pd

In [2]:
df = pd.read_csv(r"C:\Users\Hp\Downloads\churn.csv")

In [3]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
df.drop(columns = ['RowNumber','CustomerId','Surname'],inplace=True)

In [5]:
df.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  int64  
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(2), int64(7), object(2)
memory usage: 859.5+ KB


In [7]:
df.duplicated().sum()

0

In [8]:
df['Geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

In [9]:
df['Gender'].value_counts()

Male      5457
Female    4543
Name: Gender, dtype: int64

In [10]:
df['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

In [11]:
#handling imbalance data (UNDER SAMPLING)

#using under sampling( in tis process we will be taking all the points from category having less counts and we will randomly sample same number of counts from the category having high value counts
example- Excited column= 7964 (0's), 2037 (1's) so in this we will have all values of 1 and we will randomly sample 2037 values from 0. it should only be peform when we have huge number of data set

In [12]:
X = df.drop(columns=['Exited'])
y = df['Exited'].values

In [13]:
#Converting categorucal variables to numerical

In [14]:
from sklearn.preprocessing import OneHotEncoder
one_hot_encoder = OneHotEncoder(drop='first', sparse=False)
geo = one_hot_encoder.fit_transform(df[['Geography']])
gender = one_hot_encoder.fit_transform(df[['Gender']])

C:\Users\Hp\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [15]:
X_rem = X.drop(columns=['Geography','Gender'])
#X_rem

In [16]:
X_new = np.concatenate((X_rem,geo,gender),axis=1)

In [17]:
X_new.shape

(10000, 11)

# handling imbalance data (UNDER SAMPLING)

#using under sampling( in tis process we will be taking all the points from category having less counts and we will randomly sample same number of counts from the category having high value counts
example- Excited column= 7964 (0's), 2037 (1's) so in this we will have all values of 1 and we will randomly sample 2037 values from 0. it should only be peform when we have huge number of data set

In [18]:
no = df[df['Exited']==0]
yes = df[df['Exited']==1]

In [19]:
print(no.shape,yes.shape)

(7963, 11) (2037, 11)


In [20]:
#!pip install imblearn
from imblearn.under_sampling import RandomUnderSampler

In [21]:
#implementing undersampling
rs = RandomUnderSampler()

In [22]:
X_res,y_res =rs.fit_resample(X_new,y)

In [23]:
X_res.shape,y_res.shape

((4074, 11), (4074,))

In [24]:
from collections import Counter
print('Originol dataset shape {}'.format(Counter(y)))
print('resample dataset shape {}'.format(Counter(y_res)))

Originol dataset shape Counter({0: 7963, 1: 2037})
resample dataset shape Counter({0: 2037, 1: 2037})


# handling imbalanced data (OVER SAMPLING)

in oversampling we create a new data set having similar number of highest values and for less count category we increase that points in same dimension 
example - in exited = 0's are 7963 and 1's are 2037 so we will have same number of 0;s in new data set where as 1's will be increased to maintain the ratio 1:1

In [25]:
from imblearn.combine import SMOTETomek

In [26]:
smk = SMOTETomek(random_state=42)


In [27]:
X_upsam,y_upsam = smk.fit_resample(X_new,y)

In [28]:
X_upsam.shape, y_upsam.shape

((13778, 11), (13778,))

In [29]:
from collections import Counter
print('Originol dataset shape {}'.format(Counter(y)))
print('resample dataset shape {}'.format(Counter(y_upsam)))

Originol dataset shape Counter({0: 7963, 1: 2037})
resample dataset shape Counter({1: 6889, 0: 6889})


# Model Building

In [30]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_upsam,y_upsam,test_size=0.2,random_state=0)

In [31]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

X_train_trf = scaler.fit_transform(X_train)
X_test_trf = scaler.transform(X_test)

In [32]:
#pip install --upgrade tensorflow

In [33]:
import tensorflow as tf

In [35]:
from tensorflow import keras
from tensorflow.keras import Sequential 
from tensorflow.keras.layers import Dense

In [37]:
model = Sequential()

model.add(Dense(11,activation='relu',input_dim=11))
model.add(Dense(11,activation='relu'))
model.add(Dense(1,activation='sigmoid'))

In [38]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 11)                132       
                                                                 
 dense_4 (Dense)             (None, 11)                132       
                                                                 
 dense_5 (Dense)             (None, 1)                 12        
                                                                 
Total params: 276 (1.08 KB)
Trainable params: 276 (1.08 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [39]:
model.compile(optimizer='Adam',loss='binary_crossentropy',metrics=['accuracy'])

In [40]:
model.fit(X_train,y_train,batch_size=50,epochs=100,verbose=1,validation_split=0.2)

Epoch 1/100


177/177 [==============================] - 5s 12ms/step - loss: 781.1871 - accuracy: 0.5171 - val_loss: 112.2088 - val_accuracy: 0.5007
Epoch 2/100
177/177 [==============================] - 1s 7ms/step - loss: 74.1832 - accuracy: 0.5126 - val_loss: 81.2028 - val_accuracy: 0.5116
Epoch 3/100
177/177 [==============================] - 1s 8ms/step - loss: 61.0785 - accuracy: 0.5165 - val_loss: 68.9149 - val_accuracy: 0.5769
Epoch 4/100
177/177 [==============================] - 1s 8ms/step - loss: 56.5679 - accuracy: 0.5261 - val_loss: 40.4002 - val_accuracy: 0.5478
Epoch 5/100
177/177 [==============================] - 1s 7ms/step - loss: 47.7257 - accuracy: 0.5381 - val_loss: 40.3243 - val_accuracy: 0.5556
Epoch 6/100
177/177 [==============================] - 2s 11ms/step - loss: 44.2947 - accuracy: 0.5265 - val_loss: 22.9189 - val_accuracy: 0.5297
Epoch 7/100
177/177 [==============================] - 2s 10ms/step - loss: 36.1729 - accuracy: 0.5164 - val_loss: 18.0136 -

177/177 [==============================] - 1s 8ms/step - loss: 26.7932 - accuracy: 0.5446 - val_loss: 22.2300 - val_accuracy: 0.5537
Epoch 55/100
177/177 [==============================] - 1s 7ms/step - loss: 27.6573 - accuracy: 0.5537 - val_loss: 27.1340 - val_accuracy: 0.5179
Epoch 56/100
177/177 [==============================] - 1s 8ms/step - loss: 28.8248 - accuracy: 0.5542 - val_loss: 29.3050 - val_accuracy: 0.5265
Epoch 57/100
177/177 [==============================] - 1s 8ms/step - loss: 27.4253 - accuracy: 0.5446 - val_loss: 21.1990 - val_accuracy: 0.6073
Epoch 58/100
177/177 [==============================] - 1s 5ms/step - loss: 31.6281 - accuracy: 0.5445 - val_loss: 12.9275 - val_accuracy: 0.5905
Epoch 59/100
177/177 [==============================] - 1s 8ms/step - loss: 27.0768 - accuracy: 0.5577 - val_loss: 59.3280 - val_accuracy: 0.4608
Epoch 60/100
177/177 [==============================] - 1s 8ms/step - loss: 31.8368 - accuracy: 0.5522 - val_loss: 17.8492 - val_accuracy

In [41]:
y_log = model.predict(X_test_trf)

87/87 [==============================] - 1s 3ms/step


In [42]:
y_pred = np.where(y_log>0.5,1,0)

In [45]:
from sklearn.metrics import accuracy_score,classification_report
accuracy_score(y_test,y_pred)

0.636066763425254

In [46]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.67      0.60      0.64      1459
           1       0.60      0.67      0.63      1297

    accuracy                           0.64      2756
   macro avg       0.64      0.64      0.64      2756
weighted avg       0.64      0.64      0.64      2756

